core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/NativeSailStore.java

```diff
@@ -520,12 +520,18 @@ private long removeStatements(Resource subj, IRI pred, Value obj, boolean explic
 				}
 
 				final int[] contextIds = new int[contexts.length == 0 ? 1 : contexts.length];
-				if (contexts.length == 0) {
+				if (contexts.length == 0) { // remove from all contexts
 					contextIds[0] = NativeValue.UNKNOWN_ID;
 				} else {
 					for (int i = 0; i < contexts.length; i++) {
 						Resource context = contexts[i];
-						contextIds[i] = context == null ? 0 : valueStore.getID(context);
+						if (context == null) {
+							contextIds[i] = 0;
+						} else {
+							int id = valueStore.getID(context);
+							// unknown_id cannot be used (would result in removal from all contexts)
+							contextIds[i] = (id != NativeValue.UNKNOWN_ID) ? id : Integer.MIN_VALUE;
```

Comment:
can `id` be negative, if not then I propse -1 instead of MIN_VALUE

Reply:
id can indeed be negative, NativeStore.UNKNOWN_ID is -1

Which actually triggers the bug, becayse for contexts there is not distinction between not specifying the context at all (yields -1, so remove from all contexts) and a non-existing context....

Reply:
thanks for explaining :)